### PR TITLE
Reducing duplicate code by refactoring `parse_size()` of head, tail, truncate, stdbuf, etc.

### PR DESCRIPTION
**Pull request by jhscheer**
The uutils head, tail, truncate, stdbuf all have their own functions and tests
(and often individual bugs) for parsing size strings, e.g. "2KiB" or "3MB", etc.

This PR reduces duplicate code by merging the parse functions (and their tests) of:

- [x] `tail::parse_size(mut size_slice: &str) -> Result<u64, ParseSizeErr>`
- [x] `truncate::parse_size(size: &str) -> Result<u64, ()>`
- [x] `stdbuf::parse_size(size: &str) -> Option<u64>`
- [x] `head::parse::parse_num(src: &str) -> Result<(usize, bool), ParseError>`
- [x] `du::unit_string_to_number(s: &str) -> Option<u64>`
- [x] `od::parse_number_of_bytes(s: &str) -> Result<usize, &'static str>`
- [x] `sort::parse_byte_count(input: &str) -> usize`
- [x] `split::ByteSplitter::new(settings: &Settings) -> ByteSplitter`

into:
`src/uucore/src/lib/parser/parse_size::parse_size(size: &str) -> Result<usize, ParseSizeError>`

For a potential follow up PR, there is also the reverse operation, i.e. print human readable string from integer, used by df and du, etc. 

Grepping through the man pages there's also:
(dd), df, (dir), ~~du~~, ls, ~~od~~, ~~sort~~, ~~split~~, (vdir)
which could potentially also be candidates for code reduction related to 'parse_size'.

semi related:
- [x] "move uucore/features/parse_time.rs" to "uucore/parser/parse_time.rs"

### CODE REVIEW

**Comment by miDeb** [comment]
FYI: There's also `parse_byte_count` in `sort.rs`. Its behavior is a bit different in that it should default to "K" if no unit is given, and only parse the number without multiplying if the unit "B" is specified.

**Comment by tertsdiepraam** [comment]
This is awesome!

> Grepping through the man pages there's also:
(dd), df, (dir), du, ls, od, sort, split, (vdir)
which could potentially also be candidates for code reduction related to 'parse_size'.

There is no current implementation in `ls`, but it is indeed necessary for the `--block-size` argument.

> For a potential follow up PR, there is also the reverse operation, i.e. print human readable string from integer, used by df and du, etc.

This is implemented in `ls` through the functions `display_size` and `format_prefixed` functions with the [`number_prefix`](https://crates.io/crates/number_prefix) crate. You could base the `uucore` implementation on that if you want.

File: src/uu/du/src/du.rs [hunk 1]
```diff
@@ -744,31 +706,27 @@ Try '{} --help' for more information.",
     0
 }
 
+fn format_error_message(error: ParseSizeError, s: &str, option: &str) -> String {
+    // NOTE:
+    // GNU's du echos affected flag, -B or --block-size (-t or --threshold), depending user's selection
+    // GNU's du distinguishs between "invalid (suffix in) argument"
```
Comment (sylvestre) [inline src/uu/du/src/du.rs:712]:
```suggestion
    // GNU's du distinguish between "invalid (suffix in) argument"
```

File: src/uu/od/src/od.rs [hunk 2]
```diff
@@ -636,3 +639,13 @@ fn open_input_peek_reader(
     let pr = PartialReader::new(mf, skip_bytes, read_bytes);
     PeekReader::new(pr)
 }
+
+fn format_error_message(error: ParseSizeError, s: &str, option: &str) -> String {
+    // NOTE:
+    // GNU's od echos affected flag, -N or --read-bytes (-j or --skip-bytes, etc.), depending user's selection
+    // GNU's od distinguishs between "invalid (suffix in) argument"
```
Comment (sylvestre) [inline src/uu/od/src/od.rs:646]:
```suggestion
    // GNU's od distinguish between "invalid (suffix in) argument"
```

File: src/uu/sort/src/sort.rs [hunk 3]
```diff
@@ -1560,6 +1564,16 @@ fn open(path: impl AsRef<OsStr>) -> Box<dyn Read + Send> {
     }
 }
 
+fn format_error_message(error: ParseSizeError, s: &str, option: &str) -> String {
+    // NOTE:
+    // GNU's sort echos affected flag, -S or --buffer-size, depending user's selection
+    // GNU's sort distinguishs between "invalid (suffix in) argument"
```
Comment (sylvestre) [inline src/uu/sort/src/sort.rs:1570]:
```suggestion
    // GNU's sort distinguish between "invalid (suffix in) argument"
```

File: src/uucore/src/lib/lib.rs [hunk 4]
```diff
@@ -19,10 +19,10 @@ pub extern crate winapi;
 
 //## internal modules
 
-mod macros; // crate macros (macro_rules-type; exported to `crate::...`)
-
 mod features; // feature-gated code modules
+mod macros; // crate macros (macro_rules-type; exported to `crate::...`)
 mod mods; // core cross-platform modules
+mod parser; // string parsing moduls
```
Comment (sylvestre) [inline src/uucore/src/lib/lib.rs:25]:
```suggestion
mod parser; // string parsing modules
```

File: src/uu/du/src/du.rs (hunk 1 above)
Comment (jhscheer) [reply src/uu/du/src/du.rs]: done

File: src/uu/od/src/od.rs (hunk 2 above)
Comment (jhscheer) [reply src/uu/od/src/od.rs]: done

File: src/uu/sort/src/sort.rs (hunk 3 above)
Comment (jhscheer) [reply src/uu/sort/src/sort.rs]: done

File: src/uucore/src/lib/lib.rs (hunk 4 above)
Comment (jhscheer) [reply src/uucore/src/lib/lib.rs]: done

**Comment by sylvestre** [comment]
@jhscheer impressive work, is it ready?

**Comment by jhscheer** [comment]
@sylvestre thanks :)
Yes, it's ready.

File: src/uu/du/src/du.rs [hunk 5]
```diff
@@ -223,64 +227,24 @@ fn get_file_info(path: &PathBuf) -> Option<FileInfo> {
...
-
-            for env_var in &["DU_BLOCK_SIZE", "BLOCK_SIZE", "BLOCKSIZE"] {
-                let env_size = env::var(env_var).ok();
-                if let Some(quantity) = translate_to_pure_number(&env_size.as_deref()) {
-                    return quantity;
+fn read_block_size(s: Option<&str>) -> usize {
+    if let Some(s) = s {
+        parse_size(s).map_or_else(
+            |e| crash!(1, "{}", format_error_message(e, s, options::BLOCK_SIZE)),
+            |n| n,
```
Comment (tertsdiepraam) [inline src/uu/du/src/du.rs:234]: Nit, but you could use `unwrap_or_else` here.

File: src/uu/du/src/du.rs [hunk 6]
```diff
@@ -608,7 +572,7 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
         }
     };
 
-    let block_size = read_block_size(matches.value_of(options::BLOCK_SIZE));
+    let block_size = u64::try_from(read_block_size(matches.value_of(options::BLOCK_SIZE))).unwrap();
```
Comment (tertsdiepraam) [inline src/uu/du/src/du.rs:575]: There are a lot of `usize -> u64` conversions. What was your reason for using `usize` and not `u64`?

File: src/uu/truncate/src/truncate.rs [hunk 7]
```diff
@@ -290,7 +293,7 @@ fn truncate(
         }
         (Some(rfilename), None) => truncate_reference_file_only(&rfilename, filenames, create),
         (None, Some(size_string)) => truncate_size_only(&size_string, filenames, create),
-        (None, None) => crash!(1, "you must specify either --reference or --size"),
+        (None, None) => crash!(1, "you must specify either --reference or --size"), // this case cannot happen anymore because it's handled by clap
```
Comment (tertsdiepraam) [inline src/uu/truncate/src/truncate.rs:296]: I've put `unreachable!` in these situations, but this also works just fine.

File: src/uucore/src/lib/parser/parse_size.rs [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+        // has to be handled in the respective uutils because strings differ, e.g.:
+        //
+        // head:     invalid number of bytes: ‘1Y’: Value too large for defined data type
+        // tail:     invalid number of bytes: ‘1Y’: Value too large for defined data type
+        // split:    invalid number of bytes: ‘1Y’: Value too large for defined data type
+        // truncate:          Invalid number: ‘1Y’: Value too large for defined data type
+        // stdbuf:               invalid mode ‘1Y’: Value too large for defined data type
+        // sort:     -S argument '1Y' too large
+        // du:       -B argument '1Y' too large
+        // od:       -N argument '1Y' too large
```
Comment (tertsdiepraam) [inline src/uucore/src/lib/parser/parse_size.rs:157]: Wow, impressive work tracking this all down! This is definitely something that I have to keep in mind for the error handling PR.

**Review by tertsdiepraam — COMMENTED** [review]
Awesome! Very impressive! And the small improvements in other parts are great too. I'm also looking forward to integrating the `ParseError` with the error handling improvements I've been working on. I'm just wondering why you used `usize` instead of `u64` when most utils expect `u64`?

File: src/uu/du/src/du.rs (hunk 5 above)
Comment (jhscheer) [reply src/uu/du/src/du.rs:234]: But with `unwrap_or_else` I couldn't use `e` or am I mistaken?

Comment (tertsdiepraam) [reply src/uu/du/src/du.rs:234]: It takes a `FnOnce(e) -> T` on `Result`, so I think you can!

File: src/uucore/src/lib/parser/parse_size.rs (hunk 8 above)
Comment (jhscheer) [reply src/uucore/src/lib/parser/parse_size.rs:157]: Yeah, I was surprised at first that they don't all behave the same. But I guess it makes sense if you take the historic aspect into account. My conclusion is that nothing can be taken for granted and every assumption needs to be double checked.

**Comment by tertsdiepraam** [comment]
> Do you happen to know how GNU coreutils behaves if compiled for 32bit?

Nope, sorry. Although since it's C, I would guess it's dependent on the platform.

I'm not sure if this debate has been settled sometime already, but I'm wondering whether we should inherit the limitations of GNU. As in, if we can support `u64` sizes without a performance hit, why not? Here is one other example of us being slightly more lenient that I'm aware of:
```rust
$ # GNU
$ ls -w=100
ls: invalid line width: ‘=100’
$ # uutils
$ ls -w=100
<succeeds>
```
(of course we also support the valid GNU formats `-w100` and `-w 100`)

In my opinion, these are places where we could differentiate uutils from GNU, because if we do a lot of these improvements, they start to add up. But that's just my opinion of course and following GNU closely might also be a valid goal, since there is probably always somebody out there who relies on some limitation of GNU.

Anyway, all of this is not really important for this PR and `usize` makes sense in this case! :)

File: src/uu/du/src/du.rs (hunk 5 above)
Comment (jhscheer) [reply src/uu/du/src/du.rs]: Thanks for mentioning this. I wasn't aware `unwrap_or_else` could be used in this case. Now it looks much cleaner. (I feel like this should be a clippy lint.)

**Comment by jhscheer** [comment]
I'm investigating why tests from the recently merged #2364 are broken on this PR.

**Comment by miDeb** [comment]
Sorry, that's a bug with my test because I used a uppercase `B` instead of a lowercase one... Feel free to correct it.

**Review by tertsdiepraam — APPROVED** [review]
Seems like there are some conflicting files. Could you take a look at those? I'll try to prioritize this PR afterwards to it won't conflict again.

**Pull request merged**